### PR TITLE
salt: Make salt module compatible with Python3

### PR DESCRIPTION
**Pull request by TeddyAndrieux**
**Component**:

'salt'

**Context**: 

We will migrate Salt to python3 in development/2.7 but first, let's make salt modules compatible with python2 and python3

**Summary**:

Make all custom salt modules, states, ... compatible with Python3 using
six:
- replace `basestring` with `six.string_types`
- replace `urlparse` with `six.moves.urllib.parse`
- do not try to decode string object
- do not convert string to unicode
- do not use `yum` Python API as it's only available with Python2
- encode and decode string object when using `base64`

NOTE: We do this change before migrating to Python3, so that future
downgrade will be simpler

---

Refs: #2203

### CODE REVIEW

File: salt/_modules/metalk8s_volumes.py [hunk 1]
```diff
@@ -295,7 +296,7 @@ def prepare(self, force=False):
                             .format(self.path))
         storage_class = self.get('spec.storageClass')
         # If we got a string that means the name wasn't replaced by the object.
-        if isinstance(storage_class, basestring):
+        if isinstance(storage_class, six.string_types):
```
Comment (slaperche-scality) [inline salt/_modules/metalk8s_volumes.py:297]: What about l.222-223 of this file?

```python
    # XXX: Will need to be updated when moving to Python 3
    __metaclass__ = abc.ABCMeta
```
:stuck_out_tongue: 

You'll probably want https://stackoverflow.com/a/35673504

Comment (TeddyAndrieux) [reply salt/_modules/metalk8s_volumes.py:297]: Oh ! Right on my py3 test didn't faced any issue about this one. Anyway I will change this one

File: salt/metalk8s/orchestrate/downgrade/precheck.sls [hunk 2]
```diff
@@ -2,8 +2,7 @@
 
 {#- When downgrading saltenv should be the newest version #}
 {%- set nodes_versions = pillar.metalk8s.nodes.values() | map(attribute='version') | list %}
-{%- do nodes_versions.sort(cmp=salt.pkg.version_cmp, reverse=True) %}
-{%- set expected = nodes_versions | first %}
+{%- set expected = salt.metalk8s.cmp_sorted(nodes_versions, cmp=salt.pkg.version_cmp, reverse=True) | first %}
```
Comment (gdemonet) [inline salt/metalk8s/orchestrate/downgrade/precheck.sls]: Is this related to Python3 compat?

**Review by gdemonet — COMMENTED** [review]
Didn't see any issue in the diff, but didn't check the rest of the code... Do you need me to test it with Python 3?

File: salt/metalk8s/orchestrate/downgrade/precheck.sls (hunk 2 above)
Comment (gdemonet) [reply salt/metalk8s/orchestrate/downgrade/precheck.sls]: Moreover, I don't have this `cmp_sorted` method available in `salt/_modules/metalk8s.py` :confused:

Comment (TeddyAndrieux) [reply salt/metalk8s/orchestrate/downgrade/precheck.sls]: Oh shit, you right it's related to Python3 migration but not to Python3 compat, it shouldn't be in this PR. I will remove it

**Comment by TeddyAndrieux** [comment]
> Didn't see any issue in the diff, but didn't check the rest of the code... Do you need me to test it with Python 3?

You can test if you want but I already did some upgrade and downgrade test from Py2 to Py3 Salt and it worked so ... should be ok :smiley:

**Review by gdemonet — APPROVED** [review]
> You can test if you want but I already did some upgrade and downgrade test from Py2 to Py3 Salt and it worked so ... should be ok :smiley: 

Then let's go! :shipit:
